File: node-bindings/classic-node/__test__/message_logging_conformance.ts
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import * as classic from "../index";

/** Capture actual native records from a fresh process with an explicitly initialized logger. */
export function observeMessageLogging(fixture: Record<string, any>): Record<string, any> {
  const child = spawnSync(process.execPath, [fileURLToPath(import.meta.url), "--worker", JSON.stringify(fixture)], {
    encoding: "utf8", env: { ...process.env, RUST_LOG: "trace", RUST_LOG_STYLE: "never" },
  });
  if (child.status !== 0) throw new Error(`Native logging worker failed: ${child.stderr}`);
  const records = child.stderr.trim().split(/\r?\n/).filter(Boolean).map(line => {
    const match = /^\[[^\]]+\s(TRACE|DEBUG|INFO|WARN|ERROR)\s+[^\]]+\] (.*)$/.exec(line);
    if (!match) throw new Error(`Unrecognized native log record: ${line}`);
    return { level: match[1], message: match[2] };
  });
  return { records, ...JSON.parse(child.stdout) };
}

if (process.argv[2] === "--worker") {
  const fixture = JSON.parse(process.argv[3]);
  if (fixture.operation !== "basic") throw new Error("Unsupported Node logging operation");
  classic.initLogging();
  classic.initLogging();
  const logger = new classic.JsLogger(fixture.name);
  const factory = classic.createLogger(fixture.name);
  if (logger.name !== fixture.name || factory.name !== logger.name) throw new Error("Logger names disagree");
  logger.info(fixture.messages.info);
  factory.warning(fixture.messages.warning);
  logger.error(fixture.messages.error);
  factory.debug(fixture.messages.debug);
  process.stdout.write("{}");
}

File: python-bindings/tests/message_logging_conformance.py
"""Capture actual native logger records from an isolated initialized Python process."""

import json
import os
import re
import subprocess
import sys
from pathlib import Path


def _worker(fixture):
    """Invoke only public binding APIs; the parent owns the independent log observation."""
    import classic_message as message

    message.init_logging()
    message.init_logging()
    logger = message.Logger()
    operation = fixture["operation"]
    if operation == "basic":
        for level, content in fixture["messages"].items():
            getattr(logger, level)(content)
        return {}
    if operation == "format":
        return {
            "formatted": message.format_contract_event(
                **{k: v for k, v in fixture.items() if k != "operation"}
            )
        }
    if operation != "extended":
        raise ValueError("unsupported Python logging operation")
    logger.trace(fixture["trace"])
    logger.log("info", fixture["dynamic"])
    native = message.Message(fixture["message"], message.MessageType.Warning)
    native.set_title(fixture["title"])
    native.set_details(fixture["details"])
    logger.log_message(native)
    rejected = []
    for invoke in (
            lambda: logger.log(fixture["invalid"], "ignored"),
            lambda: logger.is_enabled_for(fixture["invalid"]),
    ):
        try:
            invoke()
        except ValueError:
            rejected.append(True)
        else:
            rejected.append(False)
    return {
        "enabled": [
            logger.is_info_enabled(),
            logger.is_debug_enabled(),
            logger.is_trace_enabled(),
            logger.is_enabled_for("warning"),
        ],
        "name": logger.name(),
        "invalidLog": rejected[0],
        "invalidEnabled": rejected[1],
    }


def observe_message_logging(fixture):
    """Capture severity and message bytes, omitting environmental timestamp and log target prefixes."""
    child = subprocess.run(
        [sys.executable, str(Path(__file__).resolve()), json.dumps(fixture)],
        env={**os.environ, "RUST_LOG": "trace", "RUST_LOG_STYLE": "never"},
        capture_output=True,
        text=True,
        encoding="utf-8",
        check=False,
    )
    if child.returncode != 0:
        raise ValueError(f"native logging worker failed: {child.stderr}")
    records = []
    for line in child.stderr.splitlines():
        match = re.fullmatch(
            r"\[[^\]]+\s(TRACE|DEBUG|INFO|WARN|ERROR)\s+[^\]]+\] (.*)", line
        )
        if match is None:
            raise ValueError(f"unrecognized native log record: {line}")
        records.append({"level": match[1], "message": match[2]})
    return {"records": records, **json.loads(child.stdout)}


if __name__ == "__main__":
    print(json.dumps(_worker(json.loads(sys.argv[1]))))
